### PR TITLE
Attacking shocker with worn non-conductive gloves won't zapback you

### LINKED ISSUE #37237: Swimmer Glove Spear Zap-Back
# Describe the bug

When using a reach weapon (sharpened rebar)with Swimmer Gloves, which are non-conductive, I received zap-back.

# Steps To Reproduce

1.Equip nonconductive gloves and wield a conductive weapon(possibly reach only)
2.Attack(possibly reach attack only) a enemy with zap-back

# Expected behavior

The non-conductive gloves should protect against zapback.

### PR DESCRIPTION
**Pull request by Night-Pryanik**
#### Summary
```SUMMARY: Content "Attacking shocker with worn non-conductive gloves won't zapback you."```

#### Purpose of change
Closes #37237 - Swimmer Glove Spear Zap-Back.

#### Describe the solution
Added a check in `zapback` monster defense function: if an item covers your hands and is made of non-conductive material or has `NONCONDUCTIVE` flag, you won't get zapback while attacking shockers with your (almost) bare hands or conductive weapon.

#### Describe alternatives you've considered
None.

#### Testing
Spawned swimmer gloves, wore them. Spawned shocker zombie. Attacked it with my (almost) bare hands and with conductive melee weapon. No zapback.

#### Additional context
None.

### CODE REVIEW

File: src/mondefense.cpp [hunk 1]
```diff
@@ -47,6 +47,13 @@ void mdefense::zapback( monster &m, Creature *const source,
 
     const player *const foe = dynamic_cast<player *>( source );
 
+    // Players/NPCs can avoid the shock if they wear non-conductive gear on their hands
+    for( const item &i : foe->worn ) {
```
Comment (anothersimulacrum) [inline src/mondefense.cpp]: What happens if `foe` is `nullptr` (a monster is attacking the shocker, not a player)? I think this should be under the `foe != nullptr` check.

Comment (Night-Pryanik) [reply src/mondefense.cpp]: Forgot about that. Thanks.

**Comment by ZhilkinSerg** [comment]
You would probably need to update tests - we have some for zapback.

**Comment by Night-Pryanik** [comment]
Hmm, that's strange. Tests seem to fail on `zapback_monster` case. But `if( foe != nullptr )` in my changes effectively ignores monsters. 

Also, I manually tested monster-vs-shocker test case. Spawned debug monster (gave him zapback ability in json), spawned ordinary friendly zombie. Zombie attacked debug monster, got shocked and lost some of its health.
изображение
изображение

**Comment by Grubleafeater** [comment]
Also, test reach weapons that are conductive in case it works differently, maybe?

**Comment by RarkGrames** [comment]
Does this take coverage into account? I.e. bracelets, wristwatches, fingerless gloves?

**Comment by Night-Pryanik** [comment]
> Also, test reach weapons that are conductive in case it works differently, maybe?

Tested. It works as it should.

**Comment by Night-Pryanik** [comment]
> Does this take coverage into account? I.e. bracelets, wristwatches, fingerless gloves?

No. Bracelets doesn't have any occupied body part, so they won't affect player's conductivity. All watches in the game are made from conductive materials, so they won't help either. Fingerless gloves are probably the only example which we should care about.
I thought about taking coverage into account, but couldn't decide how it should be implemented. Require 100% coverage for worn non-conductive hand gear for it to protect against electricity?

**Comment by anothersimulacrum** [comment]
I think 100% coverage would be a good way to handle it.

**Comment by RarkGrames** [comment]
95% (there is a small portion of skin showing above the wrist) or 100% coverage would be good. Also, there are some outliers, like plastic shopping bag, or game watch that would work in this case. New items can be introduced later that would break this thing.

**Comment by daydreamer0000** [comment]
What about the martial art attacks listed as kicks? Or the bonus attacks from mutations?

**Comment by Night-Pryanik** [comment]
> What about the martial art attacks listed as kicks? Or the bonus attacks from mutations?

They weren't accounted for at all prior to this PR, e.g. if you have bare foots and performing a kick while having non-conductive weapon in hands, you ain't electrocuted. Fixing this requires creating a new complex system, and this is out of the scope of this PR.

**Pull request merged**